### DISCUSSION TITLE
Question regarding Silt.NET.Windowing multi-threaded behavior?

### DISCUSSION THREAD

**Discussion by XiaochaoYan** [dotnet/Silk.NET#2256] [Q&A]
/// <summary>
        /// Calls the Render event. On a multi-threaded window, this will run on an internally managed thread.
        /// </summary>
        void DoRender();

        /// <summary>
        /// Calls the Update event. On a multi-threaded window, this can run on any thread.
        /// </summary>
        void DoUpdate();

This is what I found in src/Windowing/Silk.NET.Windowing.Common/Interfaces/IView.cs . Does this mean we can assume Silk.NET itself will create thread to boost performance when it is possbile or do we need to manually start a seperate render thread for Silk.NET?

I am also curious about the default refresh rate and v-sync related behavior.

**Comment by Perksey** [comment]
This functionality was removed in 2.0 and I guess the doc comment wasn't updated.

**Reply by XiaochaoYan** [reply]
Apologize in advance for asking an impatient question, but is there any schedule on when will documentation or any similiar support material be online? I mean clearly Silk.NET team are working hard but I can hardly say without documentaion it's near for actual production usage.

**Reply by Perksey** [reply]
No.

**Comment by cyraid** [comment]
I have successfully implemented a render thread approach and put DoRender() and DoUpdate() in it, to solve the hanging while dragging a window deal. Works pretty nice.

**Reply by Perksey** [reply]
Nice work!

**Reply by CeSun** [reply]
Hi, is there any example I can refer to?

**Reply by cyraid** [reply]
@CeSun I can give you some pointers.

Declare these in a class (everything just public for now as an example), you can call it `Application` if you like:

```csharp
  public IView? view;
  public IInputContext? inputContext;
  public Thread? renderThread; // Used for checking the current thread against the Render Thread (see more comments later).
  public readonly Thread mainThread = Thread.CurrentThread; // Can be a static, as it's only ever set once immediately in the Main Thread.
  public readonly ManualResetEventSlim isClosing = new(false); // Events are a nice way for threads to communicate, ManualReset Event means it won't automatically set back to 'untriggered' when it gets triggered.
  public readonly ManualResetEventSlim endRendering = new(false); // This is used to let other threads know the Rendering Loop is finished.
  public readonly ConcurrentQueue<Action> renderThreadQueue = new();  // Used for calling methods from the Main Thread to run the Render Thread (see `RunRenderLoop()`).
```

In your main thread (so no Task.Run's yet, etc):

```csharp
  public void Init() {

    // I tell Silk.NET not to assume the usual glfw/sdl checks as I'm know I'm using SDL only here, this section is optional //

    Window.ShouldLoadFirstPartyPlatforms(false);
    Silk.NET.Windowing.Sdl.SdlWindowing.RegisterPlatform();
    Silk.NET.Windowing.Sdl.SdlWindowing.Use();
    Silk.NET.Input.Sdl.SdlInput.RegisterPlatform();
    Silk.NET.Input.Sdl.SdlInput.Use();

    // Set the Window Options //

    var windowOptions = WindowOptions.Default with {
      Size = new Vector2D<int>(800, 600), // Window Width/Height
      Title = "Hello World", // Title of Window
      API = new GraphicsAPI(ContextAPI.OpenGLES, new APIVersion(3, 0)) // You don't have to use OpenGLES here, but I do for SkiaSharp, see below
    };

    view = Window.Create(windowOptions);

    /*
     * `IsEventDriven` when true, waits till Events arrive in view.DoEvents(),
     *   since it's in the Main Loop, and the Main Loop just waits for Events.
     *   The Render Loop is the one always going. If you've got other stuff in
     *   your main loop you don't need this set to true. If it were false you
     *   could do things like having a Main Thread ConcurrentQueue and execute
     *   methods like in `RunRenderLoop()`.
     */

    view.ShouldSwapAutomatically = true;
    view.IsEventDriven = true;

    view.Load += OnLoad;
    view.Closing += OnClose;
    view.Update += OnUpdate;
    view.Render += OnRender;
    view.FramebufferResize += OnFrameBufferResize;

    view.Initialize(); // This will call our `OnLoad()`.

    inputContext = view.CreateInput();

    foreach (var kb in inputContext.Keyboards) {
      kb.KeyDown += OnKeyDown;
    } // ForEach //

  } // Procedure //
```

In `OnLoad()` event, I create the OpenGLES context (I do this 'cause I'm using SkiaSharp, so I can work with their ANGLE on Windows platforms):

```csharp
  public void OnLoad() {

    if (view == null)
      return;

    view.CreateOpenGLES(); // If you're not doing OpenGLES (which we set in the API above), you don't have to do CreateOpenGLES
    view.ClearContext(); // <--- This is VERY important! Trust me. Otherwise event setting view.MakeCurrent() will complain there's already a context, and won't replace it. To @Perksey would love if there was at least an option to replace here. :)

  } // Procedure //
```

Have yourself Main Loop and Run Loop methods:

```csharp
  public void RunMainLoop() {

    if (view == null || Thread.CurrentThread != mainThread) // You can also check here if it's on the Main Thread if you really want.
      return;

    // We constantly see if isClosing has been triggered, which is set from OnClose() event. //

    while (!isClosing.IsSet) {
      view.DoEvents(); // Because of view.IsEventDriven true, this should pause and wait for events.
    } // While //

  } // Procedure //

  public void RunRenderLoop() {

    if (view == null)
      return;

    renderThread = Thread.CurrentThread; // We're setting the renderThread to whichever thread this is, which should be a different one. This is important for guards later on.
    view.MakeCurrent(); // Here we set the current rendering context for the current thread we're in (important for multi-threaded rendering). Just remember that you have to CLEAR the context from the main thread first. (Which is what we do before in `OnLoad()`).

    InitGraphics();

    while (!isClosing.IsSet) {

      // Here we get any methods queued up wanting to run Render Thread and execute them. //

      foreach (var action in renderThreadQueue.ToList()) {
        action();
      } // ForEach //

      // Do the DoUpdate() and DoRender() in our Render Loop. //

      view.DoUpdate();
      view.DoRender();

    } // While //

    endRendering.Set(); // Here we trigger the ManualEvent we have defined so that those that wait on it before they shut things down know the Rendering Loop is finished.

  } // Procedure //
```

That was the main course, now to finish off some of our other events:

```csharp
  public void OnClose() {
    isClosing.Set(); // Here we trigger the ManualEvent to let those waiting on it know the window is closing.
    view?.ContinueEvents(); // We wake up the Main Thread here 'cause we'll be done soon.
  } // Procedure //

  public void OnFrameBufferResize(Vector2D<int> size) {
    // The Events will be called from the Main Thread, so to avoid crashes (important!), we just enqueue our method in the Render Thread. //
    renderThreadQueue.Enqueue(() => {
      RenderDoResize(size.X, size.Y);
    });
  } // Procedure //

  public void Render(double deltaTime) {

    if (Thread.CurrentThread != renderThread)
      return;

    // Do your Render stuff here, it'll be called when `DoRender()` is called from the Render Thread. //

  } // Procedure //
```

And finally some other methods that we rely on:

```csharp
  public void InitializeGraphics() {

    if (Thread.CurrentThread != renderThread)
      return; // You could throw exception or assert here.

    // Do your graphics initialization here, it'll be run in the Render Thread. //

  } // Procedure //

  public void RenderDoResize(int width, int height) {

    // Check if we're in the Render Thread, since it's a Render-Thread needed function, it's nice to name the methods like `RenderSomeMethod()` so you can remind yourself that it's meant to be called in the Render Thread //

    if (Thread.CurrentThread != renderThread)
      return; // Can raise exception or assert here too if you want.

    // Do your Resize stuff here, like resizing frame buffer, etc.

  } // Procedure //
```

And this is what my main kinda looks like:

```csharp
  public void Run() {
    Init(); // Call our `Init()` stuff from the Main Thread
    Task.Run(() => RunRenderLoop()); // This is putting the render loop in a separate thread.
    RunMainLoop(); // We run the Main Loop in the Main Thread
    endRendering.WaitHandle.WaitOne(); // We wait for the Rendering Loop to finish before ending completely. When window is closed, it will call `OnClose()` and in there we set the `isClosing`, which the render loop constantly checks for.
  } // Procedure //
```

And somewhere in Program.cs you could do:

```csharp
var app = new Application(); // whatever you called your class
app.Run();
```

This isn't exactly the code I used, so if it doesn't run with just that you can let me know and I can try to help. Hopefully this helps!

**Reply by CeSun** [reply]
Thank you for spending so much time writing this code!

I'll try it now！ 😆

**Reply by CeSun** [reply]
Your way is doable, thank you again!
I have confirmed that it is not feasible on the glfw backend, and only the sdl backend works, and it works fine on windows, ubuntu22.04(wsl2.0), Android 14(MEIZU20 Pro).

Here is a simple code example for other developers to learn from: 
```csharp
using Silk.NET.OpenGL;
using Silk.NET.Windowing;
using Silk.NET.Windowing.Sdl;
using System.Drawing;

var options = WindowOptions.Default;
SdlWindowing.RegisterPlatform();
SdlWindowing.Use();

var window = Window.Create(options);

window.Load += () =>
{
    var gl = GL.GetApi(window);
    window.ClearContext();
    new Thread(() =>
    {
        window.MakeCurrent();
        int i = 0;
        while (window.IsClosing == false)
        {
            
            gl.ClearColor(Color.FromArgb(i % 256, 0, 0));
            gl.Clear(ClearBufferMask.ColorBufferBit);

            i++;
            window.SwapBuffers();
        };
    }).Start();


};

window.Run();


```

**Reply by cyraid** [reply]
I was able to get mine working with both GLFW and SDL. You'll have some issues when you wanna resize the window and do something as it'll be on the main thread. But hopefully you'll be able to extract as much as you can from my post. :) Glad it helped, and anyone else coming here!

The End Rendering Event and such is nice for when you want to clean up resources correctly when the main window is closed.

**Reply by CeSun** [reply]
Hello, sorry to bother you again, could you give me some help using angle on the desktop

**Reply by cyraid** [reply]
TL;DR -> Using OpenGLES with SDL should automatically use ANGLE when available, but will default to trying native graphics drivers EGL/GLES first.

Hi @CeSun, you can look at the above code (initializing with OpenGLES), and if you use the ANGLE packages, it should ship along libEGL.dll and libGLESv2.dll. Though SDL by default uses native GLES support from your native Graphics Drivers, if you really want SDL to force using ANGLE, you can do:

```csharp
var sdl = Sdl.GetApi();
sdl.SetHint(Sdl.HintOpenglESDriver, "1");
```

(If you do force it you can use environment variables `SDL_VIDEO_EGL_DRIVER` for libEGL and `SDL_VIDEO_GL_DRIVER` for libGLESv2 locations)

There has been some issues I've noticed with the nuget shipped ANGLE native drivers, so you can just grab a Chrome installation one. By looking a bit into it, https://github.com/dotnet/Silk.NET/issues/2024 might explain a bit why.

So for now just use the Chrome Installation drivers?

**Reply by cyraid** [reply]
@CeSun Btw, you'll probably have better luck using GLFW and ANGLE.

```csharp
var glfw = GlfwProvider.GLFW.Value;

glfw.InitHint(InitHint.AnglePlatformType, (int)AnglePlatformType.OpenGLES);
glfw.WindowHint(WindowHintContextApi.ContextCreationApi, ContextApi.EglContextApi);

Silk.NET.Windowing.Glfw.GlfwWindowing.RegisterPlatform();
Silk.NET.Windowing.Glfw.GlfwWindowing.Use();
Silk.NET.Input.Glfw.GlfwInput.RegisterPlatform();
```

Just keep in mind though, ANGLE is not thread safe, so make sure to only use one window (unless you combine rendering of all views into a single thread).
